### PR TITLE
Fhir import updates

### PR DESCRIPTION
**Pull request by gskathol**
Code for loading FHIR bundles to FHIR Server.

### CODE REVIEW

**Comment by graham-skylight** [comment]
I think this may be a little out of date - want to try a rebase on top of what's currently in the repo?

**Comment by spencer-skylight** [comment]
Rebased.

File: src/FunctionApps/python/IntakePipeline/fhir.py [hunk 1]
```diff
@@ -37,4 +42,158 @@ def read_fhir_bundles(container_url: str, container_prefix: str) -> Iterator[dic
 
 
 def upload_bundle_to_fhir_server(bundle: dict) -> None:
-    pass
+    """Upload a FHIR bundle to FHIR server"""
+    process_fhir_resource(bundle)
+
+
+def process_fhir_resource(
```
Comment (graham-skylight) [inline src/FunctionApps/python/IntakePipeline/fhir.py:49]: Can we get rid of this and call `import_to_fhir(bundle)` directly?

File: src/FunctionApps/python/IntakePipeline/fhir.py [hunk 2]
```diff
@@ -37,4 +42,158 @@ def read_fhir_bundles(container_url: str, container_prefix: str) -> Iterator[dic
...
+    The submissions may Bundles or individual FHIR resources.
+
+    See :py:func:`_ensure_bundle_batch` for details about Bundle
+    conversion and handling.
+
+    :param dict fhir_json: FHIR resource in json format.
+    :param str method: HTTP method to use (currently PUT or POST supported)
+    """
+    try:
+        token = get_access_token()
```
Comment (graham-skylight) [inline src/FunctionApps/python/IntakePipeline/fhir.py:77]: This won't work with the CDC azure (they got rid of this access method) - check `get_blob_client` with the `DefaultAzureCredential` bit

File: src/FunctionApps/python/IntakePipeline/fhir.py [hunk 3]
```diff
@@ -37,4 +42,158 @@ def read_fhir_bundles(container_url: str, container_prefix: str) -> Iterator[dic
...
+    retry_strategy = Retry(
+        total=3,
+        status_forcelist=[429, 500, 502, 503, 504],
+        allowed_methods=["HEAD", "PUT", "POST", "OPTIONS"],
+    )
+    adapter = HTTPAdapter(max_retries=retry_strategy)
+    http = requests.Session()
+    http.mount("https://", adapter)
+    http.mount("http://", adapter)
+    fhir_url = os.environ.get("FHIR_URL", "")
```
Comment (graham-skylight) [inline src/FunctionApps/python/IntakePipeline/fhir.py:93]: We're passing these things in from `__init__.py` - check the `get_required_config` function

File: src/FunctionApps/python/IntakePipeline/fhir.py [hunk 4]
```diff
@@ -37,4 +42,158 @@ def read_fhir_bundles(container_url: str, container_prefix: str) -> Iterator[dic
...
+    try:
+        resource_type = fhir_json["resourceType"]
+
+        if resource_type == "Bundle":
```
Comment (graham-skylight) [inline src/FunctionApps/python/IntakePipeline/fhir.py:97]: They're all bundles at this point - I'd vote to get rid of the other conditions to simplify things a bit

File: src/FunctionApps/python/IntakePipeline/fhir.py [hunk 5]
```diff
@@ -37,4 +42,158 @@ def read_fhir_bundles(container_url: str, container_prefix: str) -> Iterator[dic
...
+            )
+
+    except Exception:
+        logging.exception(
+            "Request using method " + method + " failed for json: " + str(fhir_json)
+        )
+        return
+
+
+def _ensure_bundle_batch(fhir_json: dict, method: str) -> dict:
```
Comment (graham-skylight) [inline src/FunctionApps/python/IntakePipeline/fhir.py:137]: I think this may be overly defensive for where we're at right now - maybe skip for now?

File: src/FunctionApps/python/tests/test_fhir.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
+import unittest
+
+from unittest import mock
+from IntakePipeline import fhir
+
+import json
+import os
+
+
+class TestFhir(unittest.TestCase):
```
Comment (graham-skylight) [inline src/FunctionApps/python/tests/test_fhir.py:10]: the unittest classes are kind of the older way to do things here - we've been on `pytest` for all of the other function apps - were you able to get that running?

tl;dr though is that these tests should be individual functions in `tests/IntakePipeline/fhir_test.py` with the other tests for that module

**Review by graham-skylight — DISMISSED** [review]
Looks like a good start, but several things won't work out with the new access control stuff

File: src/FunctionApps/python/tests/test_fhir.py (hunk 6 above)
Comment (spencer-skylight) [reply src/FunctionApps/python/tests/test_fhir.py:10]: I was able to get pytest running in my environment, and confirmed that it was picking up these tests written as they are.  I'll change them though to make them consistent with other tests.

File: src/FunctionApps/python/IntakePipeline/fhir.py (hunk 5 above)
Comment (spencer-skylight) [reply src/FunctionApps/python/IntakePipeline/fhir.py:137]: There are a few different commonly used bundle types.  I wasn't sure what they look like coming in...  I'll try to confirm what they look like - if no modifications are needed (or if those modifications can be incorporated earlier in the process), I'll remove this code.

File: src/FunctionApps/python/tests/test_fhir.py (hunk 6 above)
Comment (spencer-skylight) [reply src/FunctionApps/python/tests/test_fhir.py]: Test cases have been updated.

File: src/FunctionApps/python/IntakePipeline/fhir.py (hunk 5 above)
Comment (spencer-skylight) [reply src/FunctionApps/python/IntakePipeline/fhir.py]: The bundle modification code has been removed.  We confirmed the incoming data will be of type: "batch" and have the correct request data associated (PUT).  This seems to be how the bundles are coming back from the MS FHIR converter.  I did some basic testing locally, and the info seemed to be uploaded as expected without modifications.

File: src/FunctionApps/python/IntakePipeline/fhir.py (hunk 4 above)
Comment (spencer-skylight) [reply src/FunctionApps/python/IntakePipeline/fhir.py]: Done

File: src/FunctionApps/python/IntakePipeline/fhir.py (hunk 3 above)
Comment (spencer-skylight) [reply src/FunctionApps/python/IntakePipeline/fhir.py]: Done.

File: src/FunctionApps/python/IntakePipeline/fhir.py (hunk 2 above)
Comment (spencer-skylight) [reply src/FunctionApps/python/IntakePipeline/fhir.py]: Auth updated and tested locally using the Azure CLI in conjunction with updated python code.

File: src/FunctionApps/python/IntakePipeline/fhir.py (hunk 1 above)
Comment (spencer-skylight) [reply src/FunctionApps/python/IntakePipeline/fhir.py]: Done.

**Comment by spencer-skylight** [comment]
All comments have been addressed, and ready for re-review.

**Review by JSkeets — APPROVED** [review]
These changes LGTM!

File: src/FunctionApps/python/IntakePipeline/fhir.py [hunk 7]
```diff
@@ -46,5 +51,94 @@ def store_bundle(container_url: str, prefix: str, bundle: dict) -> None:
...
+):
+    """Import a FHIR resource to the FHIR server.
+    The submissions may Bundles or individual FHIR resources.
+
+    :param AzureFhirserverCredentialManager fhirserver_cred_manager: Credential manager.
+    :param dict fhir_json: FHIR resource in json format.
+    :param str method: HTTP method to use (currently PUT or POST supported)
+    """
+    try:
+        token = fhirserver_cred_manager.get_access_token()
```
Comment (graham-skylight) [inline src/FunctionApps/python/IntakePipeline/fhir.py:116]:
```suggestion
        token = DefaultAzureCredentials().get_token(fhir_url)
```

**Review by graham-skylight — COMMENTED** [review]
I think the credential access is overengineered, but I'm removing my objection if everyone else is good here.
